File: 백준/js/boj_14888.js
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
let input = require("fs").readFileSync(filePath).toString().trim().split("\n");

const N = Number(input[0]);
const num = input[1].split(" ").map(Number);
const op = input[2].split(" ").map(Number);
let [max, min] = [Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER];

function dfs(idx, res) {
  if (idx === N) {
    max = Math.max(max, res);
    min = Math.min(min, res);
    return;
  }

  for (let i = 0; i < 4; i++) {
    if (op[i] == 0) continue;

    op[i]--;

    switch (i) {
      case 0:
        dfs(idx + 1, res + num[idx]);
        break;
      case 1:
        dfs(idx + 1, res - num[idx]);
        break;
      case 2:
        dfs(idx + 1, res * num[idx]);
        break;
      case 3:
        if (res < 0) {
          dfs(idx + 1, -Math.floor(-res / num[idx]));
        } else {
          dfs(idx + 1, Math.floor(res / num[idx]));
        }
        break;
    }

    op[i]++;
  }
}

dfs(1, num[0]);
console.log(`${max}\n${min}`);

File: 백준/python/boj_14888.py
import sys

sys.stdin = open("input.txt", "r")

N = int(input())
num = list(map(int, input().split(" ")))
op = list(map(int, input().split(" ")))

max_num = -1e9
min_num = 1e9

def dfs(idx, res):
    global max_num
    global min_num
    if idx == N:
        max_num = max(max_num, res)
        min_num = min(min_num, res)
        return

    for i in range(4):
        if op[i]==0:
            continue

        op[i] -= 1
        if i == 0:
            dfs(idx+1, res+num[idx])
        elif i == 1:
            dfs(idx+1, res-num[idx])
        elif i == 2:
            dfs(idx+1, res*num[idx])
        elif i == 3:
            if res < 0:
                dfs(idx+1, -int(-res / num[idx]))
            else:
                dfs(idx+1, int(res / num[idx]))

        op[i] += 1

dfs(1, num[0])
print(max_num)
print(min_num)





